MigrationV3V4/CSharp/Skills/V3EchoBot/Authentication/CustomAllowedCallersClaimsValidator.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Microsoft.Bot.Connector.SkillAuthentication;
+using System;
+using System.Collections.Generic;
+using System.Security.Claims;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Sample.EchoBot.Authentication
+{
+    /// <summary>
+    /// Sample claims validator that loads an allowed list from configuration if present
+    /// and checks that requests are coming from allowed parent bots.
+    /// </summary>
+    public class CustomAllowedCallersClaimsValidator : ClaimsValidator
+    {
+        private readonly IList<string> _allowedCallers;
+
+        public CustomAllowedCallersClaimsValidator(IList<string> allowedCallers)
+        {
+            // AllowedCallers is the setting in web.config file
+            // that consists of the list of parent bot IDs that are allowed to access the skill.
+            // To add a new parent bot simply go to the AllowedCallers and add
+            // the parent bot's Microsoft app ID to the list.
+
+            _allowedCallers = allowedCallers ?? throw new ArgumentNullException(nameof(allowedCallers));
+        }
+
+        public override Task ValidateClaimsAsync(IList<Claim> claims)
+        {
+            // if _allowedCallers has one item of '*', allow all parent bot calls and do not validate the appid from claims
+            if (SkillValidation.IsSkillClaim(claims) && !(_allowedCallers.Count == 1 && _allowedCallers[0] == "*"))
+            {
+                // Check that the appId claim in the skill request is in the list of skills configured for this bot.
+                var appId = JwtTokenValidation.GetAppIdFromClaims(claims).ToUpperInvariant();
+                if (!_allowedCallers.Contains(appId))
+                {
+                    throw new UnauthorizedAccessException($"Received a request from a bot with an app ID of \"{appId}\". To enable requests from this caller, add the app ID to your configuration file.");
+                }
+            }
+
+            return Task.CompletedTask;
```

Comment:
>return Task.CompletedTask; [](start = 12, length = 26)

Why isn't flow like this?

1. if "is skill claim" and "all callers allowed", return Task.CompletedTask
2. if "is skill claim" and claim's appid matches allowed callers, return Task.CompletedTask
3. otherwie, throw unauthorized

In the current code, if there is no skill claim, then validation passes.

Reply:
Tightened this down as you've recommended.  Thanks!